Move tunnel analyzers to packet analysis framework

timwoj opened the pull request:
This is a very early draft PR for a project that I’ve been working on that kind of morphed out of control as time went on. The initial work was simply to move the tunnel analyzers from the session analysis framework to the packet analysis framework. What it ended up as:

- Moved from using individual templated types for `Tag` values to a single unified `enum` for `Tag`s. The individual types still exist for consistency, but the type definitions have been cleaned up considerably and the entire structure of `Tag`s, `Component`s, and how they’re added to `ComponentManager` is a lot neater. It also adds a unified `EnumType` for the analyzer tags called `AllAnalyzers`, which gets used in the events mentioned below.
- The `protocol_confirmation` and `protocol_violation` events are now deprecated, and replaced by new events called `analyzer_confirmation` and `analyzer_violation`. The reason for this is that the old events only worked for session analyzers. Now that the tunnel analyzers moved, they still needed support for confirmation and violation, and needed to take the `AllAnalzyers::Tag` type as an argument. Due to the way that event overloading works in Zeek, new events were needed. This also deprecates the `Analyzer::ProtocolConfirmation()` and `Analyzer::ProtocolViolation()` methods.
- Moved the tunnel analyzers to packet analyzers.
- Added the ability to forward from the UDP packet analyzer to other packet analyzers based on the port number.
- A very small change was made to `bifcl` to allow `nullptr` to be passed to enqueue methods generated from bif files. We were already doing something similar in `Session::EnqueueEvent`, so this just extends that a little to the generated methods. This makes some things in the GTP analyzer considerably easier.

Known issues:

- There’s a few TODOs still sprinkled throughout the commits. These were just notes to myself about things I didn’t quite understand at the time, and I need to go back and clean them up
- The changes to the `core.print-bpf-filters` btest indicates that we’re losing the ability to create BPF filters on the port numbers for these analyzers. It’s something I need to resolve, since that’s lost functionality for sure.

This is probably best reviewed commit-by-commit, especially the first few commits that do a ton of setup architectural changes to how tags work. These commits are technically changes for #1771, but they’re caused and needed by the rest of the tunnel work so they’re all together in one PR.

timwoj commented:
I fixed the one known problem with the BPF filters via https://github.com/zeek/zeek/pull/1838/commits/1fbf5b314a7a5728309f972b9a5041f6bb60e61f and some changes to the analyzer commits. I also rebased it on top of the current master.

I'll leave this alone for now until someone's had a chance to review it.

rsmmr commented on scripts/base/frameworks/analyzer/main.zeek:
```diff
@@ -126,6 +126,10 @@ export {
 		ANALYZER_TCPSTATS,
 	} &redef;
+
+	## Adds a port to the table of ports stored for the analyzer framework.
```
Suggest to expand this a bit explain difference to `register_for_port`.

rsmmr replied:
Also, maybe think about more description name for the function. The "to_table" part doesn't convey any semantics on what's the thing that that table is doing.

timwoj replied:
Yah, it's a tough one. How about making `register_for_port` check if an analyzer tag is a session analyzer. It could do all of the table management still, but only call `__register_for_port` if it needs to.

timwoj replied:
I think https://github.com/zeek/zeek/pull/1838/commits/0aafa1779498ce039213b5a655fb8589c2c00ce5 should resolve this.

rsmmr commented on scripts/base/packet-protocols/ayiya/main.zeek:
```diff
@@ -0,0 +1,24 @@
+const ayiya_ports = { 5072/udp };
+redef likely_server_ports += { ayiya_ports };
+
+event zeek_init() &priority=20
+	{
+	PacketAnalyzer::register_protocol_detection(PacketAnalyzer::ANALYZER_UDP, PacketAnalyzer::ANALYZER_AYIYA);
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_AYIYA, IPPROTO_IPV4, PacketAnalyzer::ANALYZER_IP);
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_AYIYA, IPPROTO_IPV6, PacketAnalyzer::ANALYZER_IP);
+
+	for ( p in ayiya_ports )
+		{
+		PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_UDP, port_to_count(p),
+		                                         PacketAnalyzer::ANALYZER_AYIYA);
+		Analyzer::add_port_to_table(PacketAnalyzer::ANALYZER_AYIYA, p);
+		}
+	}
```
I like this. This way of setting this up looks pretty straight-forward and intuitive to me.

timwoj replied:
Me too!

rsmmr commented on src/packet_analysis/protocol/ayiya/AYIYA.cc:
```diff
@@ -0,0 +1,78 @@
+	if ( ! BifConst::Tunnel::enable_ayiya )
+		return false;
+
+	if ( packet->encap && packet->encap->Depth() >= BifConst::Tunnel::max_depth )
```
Would be nice if we had some external place where this check could be done, doesn't feel like something a specific (and every) analyzer should be in charge of. That said, I can't think of a place where that would be.

timwoj replied:
Would changing it to a method like `Packet::EncapValid()` be enough?

rsmmr replied:
Yeah, that's would help, and I still don't have a better ideas where we could centralize this, so yes.

timwoj replied:
Yah I see what you mean, looking at it again. Even adding a function that does the check doesn't change the fact that we're still having to send the weird in all of those places. I'd move it up into `Analyzer::ForwardPacket`, but that doesn't feel like the right place to do it either.

timwoj replied:
We decided to leave this one alone, since the original session analyzers were doing the same thing anyways. There's not really a good place to put a function that does this either.

rsmmr commented on scripts/base/packet-protocols/teredo/main.zeek:
```diff
@@ -0,0 +1,34 @@
+module PacketAnalyzer::TEREDO;
+
+# This needs to be loaded here so the function is available. Function BIFs normally aren't
```
"so the functions are available"

rsmmr commented on src/IPAddr.cc:
```diff
@@ -24,46 +24,115 @@ namespace detail
 
 ConnKey::ConnKey(const IPAddr& src, const IPAddr& dst, uint16_t src_port, uint16_t dst_port,
```
The `ConnKey` changes come a bit surprising in this commit. What's the gist of the change there? (Hard to see from the diff without digging quite a bit).

timwoj replied:
The Teredo analyzer's `remove_teredo_analyzer` method needed a way to generate a `ConnKey` from a `Val*`, since that's what the `conn_id` type in script-land is. We already had code to do that, but it was part of `session::Manager::FindConnection()`, not part of `ConnKey` itself. The big changes here move that code out into a new constructor that takes a Val*, refactor some of the common initialization code used in all of the constructors into a single method, and then call that new constructor in the places where it's needed.

rsmmr reviewed (commented):
There's a lot going on here. :) I generally do like this, and the end result looks like a nice improvement. I made a first pass so far, but not fully done yet. Some questions:

- per some inline comments, I got confused about the relationship between analyzer_confirmation vs protocol_confirmation, and whether/how the former is replacing the latter.

- also not sure when we're using `AllAnalyzers` vs the specific/existing ones

- for the individual analyzers, it's hard to see if anything substantial has changed in their implementation; if so, can you call out the places worth looking at?

rsmmr commented on src/plugin/TaggedComponent.h:
```diff
@@ -24,7 +25,7 @@ template <class T> class TaggedComponent
 	 * If not used, leave at zero.
 	 */
-	explicit TaggedComponent(typename T::subtype_t subtype = 0);
+	explicit TaggedComponent(Tag::subtype_t subtype = 0, zeek::EnumTypePtr etype = nullptr);
```
(or change the other place)

timwoj replied:
`TaggedComponent` gets removed in later commits, so I'm just going to resolve this one.

rsmmr commented on scripts/base/frameworks/analyzer/main.zeek:
```diff
@@ -46,7 +46,7 @@ export {
 	##
 	## Returns: True if the ports were successfully registered.
-	global register_for_ports: function(tag: Analyzer::Tag, ports: set[port]) : bool;
+	global register_for_ports: function(tag: AllAnalyzers::Tag, ports: set[port]) : bool;
```
Also, a test checking typing of the tag types (individual vs aggregate) would good.

timwoj replied:
I reverted this change in https://github.com/zeek/zeek/pull/1838/commits/0aafa1779498ce039213b5a655fb8589c2c00ce5, since I added new methods for the packet analyzers.

timwoj commented on src/Tag.cc:
```diff
@@ -75,6 +77,7 @@ const EnumValPtr& Tag::AsVal() const
 	{
 	// TODO: this probably isn't valid, and we should just return the null val
 	// if it's null.
+	// TODO: should we check for a valid etype here or just let the assert fail?
```
Sounds good.

timwoj commented on src/Tag.cc:
```diff
@@ -68,11 +71,13 @@ Tag& Tag::operator=(const Tag&& other) noexcept
-const EnumValPtr& Tag::AsVal(const EnumTypePtr& etype) const
+const EnumValPtr& Tag::AsVal() const
 	{
+	// TODO: this probably isn't valid, and we should just return the null val
+	// if it's null.
```
I changed the method to just return `val` directly and everything seems fine. I'll go with that.

timwoj commented on src/iosource/Packet.h:
```diff
@@ -241,6 +248,12 @@ class Packet
 
+	/**
+	 * The session related to this packet, if one exists. This is used for
+	 * handling encapsulations, and is only set by the UDP analyzer.
```
I think that's reasonable.

timwoj commented on src/analyzer/analyzer.bif:
```diff
@@ -38,14 +38,31 @@ function Analyzer::__schedule_analyzer%(orig: addr, resp: addr, resp_p: port,
+		component = zeek::packet_mgr->Lookup(val);
+	if ( ! component )
+		component = zeek::file_mgr->Lookup(val);
+	if ( ! component )
```
The only way that could happen is if someone passed a different kind of `Tag` in here (not an analyzer tag). That should probably be an error if it was to happen though.

timwoj commented on src/plugin/ComponentManager.h:
```diff
@@ -120,20 +124,45 @@ template <class C> class ComponentManager
 	zeek::detail::add_type(id.get(), tag_enum_type, nullptr);
 	zeek::detail::zeekygen_mgr->Identifier(std::move(id));
+
+	if ( ! parent_module.empty() )
```
It's the script-level naming. We insert the tags into modules (`Analyzer`, `PacketAnalyzer`, etc). The parent module is the one that a tag can also get inserted into (in this code, `AllAnalyzers`).

timwoj commented on src/packet_analysis/Analyzer.h:
```diff
@@ -98,6 +100,28 @@ class Analyzer
+	 * matching fails. This will also be preferred over the default analyzer
+	 * if one exists.
+	 *
+	 * @param child The analyzer that will be called for protocol detection.
```
I don't think so. We don't discuss that lifetime for the registration against identifiers. Packet analyzers exist as long as zeek is running, so they should always be valid.

timwoj commented on src/packet_analysis/Analyzer.h:
```diff
@@ -98,6 +100,28 @@ class Analyzer
+
+	/**
+	 * Detects whether the protocol for an analyzer can be found in the packet
+	 * data. This can be used similarly to how DPD is used in the session
+	 * analysis tree.
```
Added comments.

timwoj commented on src/packet_analysis/protocol/iptunnel/IPTunnel.h:
```diff
@@ -78,6 +78,27 @@ class IPTunnelAnalyzer : public Analyzer
+ * Utility function for packet analyzers for tunnel protocols. This builds a new
+ * packet and sets up the encapsulation data based on the provided information.
+ *
+ * TODO: Comments here
```
The intention is just to add a method that creates the inner packet object and adds to the encapsulation stack in a single place so that the analyzers don't have to duplicate that code every time. I added comments to it, and made the method return the new packet instead of having the caller pass it in.

timwoj replied:
Ideally, I'd like to deprecate/remove the other `ProcessEncapsulatedPacket` methods, or at least make them private, since the only code using them anymore is the IP-in-IP tunnel analyzer.

timwoj commented on scripts/policy/protocols/conn/known-services.zeek:
```diff
@@ -298,6 +298,11 @@ event protocol_confirmation(c: connection, atype: Analyzer::Tag, aid: count) &pr
 	known_services_done(c);
 	}
 
+event analyzer_confirmation(c: connection, atype: AllAnalyzers::Tag, aid: count) &priority=-5
```
Yes, because if someone elsewhere is still emitting the `protocol_confirmation` events in a plugin, removing them would break things, right?

timwoj replied:
I could add something somewhere that basically just reschedules any `protocol_confirmation`/`protocol_violation` events as the `analyzer` versions, and remove all of these. It'd have a slight performance difference since the event wouldn't get handled until the next cycle, but that might be ok?

timwoj replied:
I removed the handling of the `protocol_` events from all of the base scripts, and changed the two methods in `Analyzer` to just enqueue both the old and new events.

timwoj commented on src/analyzer/Analyzer.cc:
```diff
@@ -712,6 +713,41 @@ void Analyzer::ProtocolViolation(const char* reason, const char* data, int len)
 	event_mgr.Enqueue(protocol_violation, ConnVal(), tval, val_mgr->Count(id), std::move(r));
 	}
 
+void Analyzer::AnalyzerConfirmation(zeek::Tag arg_tag)
```
See the comment on https://github.com/zeek/zeek/pull/1838#discussion_r747500584

timwoj commented:
> for the individual analyzers, it's hard to see if anything substantial has changed in their implementation; if so, can you call out the places worth looking at?

I honestly can't think of anything specific. Most of the big changes to the analyzers are moving from calls to `IPTunnel'`s `ProcessEncapsulatedPacket` methods to direct forwarding in the packet analyzer world.

rsmmr commented:
I looked over the changes addressing review feedback, that all looks good, thanks! I still want to make one more global pass now, which I haven't gotten to yet; will do soon.

rsmmr commented:
Ok, made one more pass and the changes look all good to me. 

But I noticed something with the script-side type checking. Look at this:

```
global test2: function(a: Analyzer::Tag);

function test1(a: AllAnalyzers::Tag) {
    print "all", a;
    test2(a);
}

function test2(a: Analyzer::Tag) {
    print "proto", a;
}

event zeek_init() {
    test1(Analyzer::ANALYZER_DNS);
    test2(Analyzer::ANALYZER_DNS);

    # I would expect these two to fail:
    test1(tcp);
    test2(tcp);
}
```

The first two function calls in `zeek_init()` work as expected. The next two however get accepted even though I'd think they shouldn't because it's a completely unrelated enum.

I don't recall if we have discussed this: is that a result of the existing enum type checking not working as one would expect? If so, hadn't you been trying to fix that?

Also, I think we still need a test along the lines of the code snippet above, checking that our analyzers enums can be passed around in the way we want them.

And finally, we need to document the relationship between the analyzers tag types somewhere.

timwoj commented:
> I don't recall if we have discussed this: is that a result of the existing enum type checking not working as one would expect? If so, hadn't you been trying to fix that?

Right, that's probably https://github.com/zeek/zeek/issues/1820. I ended up not fixing it because it wasn't holding up anything with the tunnel work.

timwoj commented:
> Also, I think we still need a test along the lines of the code snippet above, checking that our analyzers enums can be passed around in the way we want them.

Added in https://github.com/zeek/zeek/pull/1838/commits/2690ee32b964ba4f81984420fd325c593f811794

> And finally, we need to document the relationship between the analyzers tag types somewhere.

I added a little blurb to the Analyzer framework's main.zeek in https://github.com/zeek/zeek/pull/1838/commits/2c88634ca905ae3899cc9afde19753c243f2a6aa, but it doesn't feel like a great place for it. I'm not really sure of a better one though.

timwoj commented:
One thing that just came to mind while I was updating the documentation for the new registration methods for packet analysis is that protocol detection overrides any other methods for looking at whether a packet should be processed. For example, if protocol detection matches a packet erroneously and a violation later happens during processing, it won't attempt a simpler identifier match even if one exists. Should that be the case? See below for context: https://github.com/zeek/zeek/blob/2c88634ca905ae3899cc9afde19753c243f2a6aa/src/packet_analysis/Analyzer.cc#L116-L147

timwoj commented:
I also opened the above linked PR for the docs updates. Notably, I added a section to the packet analysis framework page about the new registration capabilities added in this PR.

rsmmr commented:
> One thing that just came to mind while I was updating the documentation for the new registration methods for packet analysis is that protocol detection overrides any other methods for looking at whether a packet should be processed.

Two thoughts:

1. With the current structure, I think that's ok. We always try only one analyzer, and for example also cannot have two analyzers for the same ID. I'd see protocol detection along the same line, it's saying: "that's me". (In my mind, this is different from DPD's two-stage confirmation, which I'm not sure is really that useful at the packet level).

2. That said, I was pondering a related question recently with Spicy: if we have two packet analyzers for the same ID, we only activate one right now, and it's not well-defined which one because it depends on registration order (symptom of this is that if Spicy re-defines a packet analyzer, that may or may not work; and one may or may not get a warning about what's happing). I've been wondering if we should change packet analyzers so that we keep a list of analyzers per ID, and forward the packet to all of them. If we did that, then I think that would solve the problem here too: we'd do protocol detection, but also forward to other analyzers.

rsmmr commented:
Test and docs look good, thanks!

I think we're good to merge!

bbannier commented on src/file_analysis/Tag.h:
```diff
@@ -6,105 +6,9 @@
-	 */
-	explicit Tag(EnumValPtr val);
-	};
+using Tag [[deprecated("Remove in v5.1. Use zeek::Tag.")]] = zeek::Tag;
```
Just FYI, this is a breaking change.

While this ensures that the type `zeek::file_analysis::Tag` can still resolve it breaks any code which overloaded functions for different `Tag` types, see e.g., [here](https://github.com/zeek/spicy-plugin/blob/2331bb0522ed104685c94036e60fb49ee43d9821/include/plugin.h#L172-L207). Maybe even worse, code which previously explicitly only matched a `file_analysis::Tag` argument can now match _any `Tag`_ argument; this could show up in e.g., visitors or other template code, or in code which uses function overloading for a `Tag` type and non-`Tag` types.

I think the correct approach instead of using an alias would have been to refactor the old type to contain a `zeek::Tag` and provides its interface, but ensure that it cannot become a `zeek::Tag` (e.g., through implicit conversions).

Here and for other `Tag` types as well.
